### PR TITLE
role model implementation

### PR DESCRIPTION
**Pull request by neorezuss**

### CODE REVIEW

**Review by evensky1 — APPROVED** [review]

File: educationapplication/src/main/resources/role_model.sql [hunk 1]
```diff
@@ -0,0 +1,67 @@
+create table if not exists users (
+     id             varchar(40)  not null primary key,
+     subscriptionId varchar(40)  not null,
+     username       varchar(255) not null,
+     first_name     varchar(255) not null,
+     last_name      varchar(255) not null,
+     email          varchar(255) not null,
```
Comment (MikleGod) [inline educationapplication/src/main/resources/role_model.sql:7]: only email should be not null (between logical fields)

File: educationapplication/src/main/resources/role_model.sql [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+create table if not exists roles (
+    id        varchar(255) not null primary key,
+    name      varchar(255) not null,
+    "system"  boolean      not null
+);
+
+create table if not exists permissions (
+    id          varchar(255) not null primary key,
+    description varchar(255) not null
```
Comment (MikleGod) [inline educationapplication/src/main/resources/role_model.sql:20]: I belive there should be 'name', or 'key' or something

Comment (MikleGod) [reply educationapplication/src/main/resources/role_model.sql:20]: Dont think that making the name as id is best option

File: modules/course-service/api/src/main/java/com/innowise/educationalsystem/course/controller/CourseController.java [hunk 3]
```diff
@@ -30,33 +31,38 @@ public class CourseController {
...
     @PutMapping
+    @PreAuthorize("hasAuthority('COURSE_EDIT')")
     public ResponseEntity<CourseDto> updateCourse(@Valid @RequestBody UpdateCourseDto updateCourseDto) {
         Course course = courseMapper.updateCourseDtoToCourse(updateCourseDto);
         Course updatedCourse = courseService.save(course);
         return ResponseEntity.ok().body(courseMapper.courseToCourseDto(updatedCourse));
     }
 
     @DeleteMapping("/{id}")
+    @PreAuthorize("hasAuthority('COURSE_EDIT')")
```
Comment (MikleGod) [inline modules/course-service/api/src/main/java/com/innowise/educationalsystem/course/controller/CourseController.java:65]: Course delete here (only Admin will be able to do so)

File: modules/urm/api/src/main/java/com/innowise/educationalsystem/controller/UserController.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    @GetMapping("/{id}")
+    @PreAuthorize("hasAuthority('USER_GET')")
+    public ResponseEntity<UserResponseDto> getById(@PathVariable("id") String id) {
+        User user = userService.getById(id);
+        return ResponseEntity.ok().body(userMapper.entityToDto(user));
+    }
+
+    @PostMapping("/{id}/assign-roles")
+    @PreAuthorize("hasAuthority('USER_ASSIGN_ROLE')")
+    public ResponseEntity<UserResponseDto> assignRoles(@PathVariable("id") String id,
```
Comment (MikleGod) [inline modules/urm/api/src/main/java/com/innowise/educationalsystem/controller/UserController.java:47]: Lets put all of the params on a separate line with 1 tab from the method declaration level

File: modules/urm/impl/src/main/java/com/innowise/educationalsystem/dto/NoSuchRoleException.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.innowise.educationalsystem.dto;
+
+public class NoSuchRoleException extends RuntimeException{
```
Comment (MikleGod) [inline modules/urm/impl/src/main/java/com/innowise/educationalsystem/dto/NoSuchRoleException.java:3]: Pls check the common exceptions lib after it is merged

File: educationapplication/src/main/resources/role_model.sql [hunk 6]
```diff
@@ -0,0 +1,67 @@
+create table if not exists users (
+     id             varchar(40)  not null primary key,
+     subscriptionId varchar(40)  not null,
+     username       varchar(255) not null,
```
[the remark this answers is not in the document]
Comment (Lelikbjmak) [reply educationapplication/src/main/resources/role_model.sql:4]: We use invitations for users. user will be created simultaneously with invitation. Actually they must be nullable. If not replaces with placeholders.

File: modules/urm/impl/src/main/java/com/innowise/educationalsystem/repository/PermissionRepository.java [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package com.innowise.educationalsystem.repository;
+
+import com.innowise.educationalsystem.entity.Permission;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface PermissionRepository extends JpaRepository<Permission, String> {
```
Comment (MikleGod) [inline modules/urm/impl/src/main/java/com/innowise/educationalsystem/repository/PermissionRepository.java:6]: Check common crud lib

File: modules/urm/impl/src/main/java/com/innowise/educationalsystem/service/RoleService.java [hunk 8]
```diff
@@ -0,0 +1,18 @@
+package com.innowise.educationalsystem.service;
+
+import com.innowise.educationalsystem.dto.RoleRequestDto;
+import com.innowise.educationalsystem.entity.Role;
+
+import java.util.List;
+
+public interface RoleService {
```
Comment (MikleGod) [inline modules/urm/impl/src/main/java/com/innowise/educationalsystem/service/RoleService.java:8]: Check common crud lib

File: modules/urm/impl/src/main/java/com/innowise/educationalsystem/service/impl/InviteServiceImpl.java [hunk 9]
```diff
@@ -36,12 +42,18 @@ public class InviteServiceImpl implements InviteService {
...
+        rejectOldInvites(inviteRequestDto.getEmail());
+
+        List<Role> roles = roleRepository.findAllById(inviteRequestDto.getRoleIds());
+        Invite invite = Invite.builder()
+                .email(inviteRequestDto.getEmail())
+                .roles(new HashSet<>(roles))
+                .subscriptionId(subscriptionId)
+                .validationExpiredAt(LocalDateTime.now().plus(validationExpirationTime))
+                .status(InviteStatus.NEW)
+                .build();
```
Comment (MikleGod) [inline modules/urm/impl/src/main/java/com/innowise/educationalsystem/service/impl/InviteServiceImpl.java:55]: ❤️

File: modules/urm/impl/src/main/java/com/innowise/educationalsystem/service/impl/RoleServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+import com.innowise.educationalsystem.service.RoleService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+import java.util.HashSet;
+import java.util.List;
+
+@Service
+@RequiredArgsConstructor
+public class RoleServiceImpl implements RoleService {
```
Comment (MikleGod) [inline modules/urm/impl/src/main/java/com/innowise/educationalsystem/service/impl/RoleServiceImpl.java:19]: Crud lib can help do less boilerplate

**Review by MikleGod — APPROVED** [review]

**Pull request merged**